### PR TITLE
[SPARK-39216][SQL] Do not collapse projects in CombineUnions if it hasCorrelatedSubquery

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

Makes `CombineUnions` do not collapse projects if it hasCorrelatedSubquery. For example:
```sql
SELECT (SELECT IF(x, 1, 0)) AS a
FROM (SELECT true) t(x)
UNION
SELECT 1 AS a
```

It will throw exception:
```
java.lang.IllegalStateException: Couldn't find x#4 in [] 
```

### Why are the changes needed?

Fix bug.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test.

### CODE REVIEW

**Comment by wangyum** [comment]
cc @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -1372,15 +1374,16 @@ object CombineUnions extends Rule[LogicalPlan] {
         // Push down projection through Union and then push pushed plan to Stack if
         // there is a Project.
         case Project(projectList, Distinct(u @ Union(children, byName, allowMissingCol)))
-            if projectList.forall(_.deterministic) && children.nonEmpty &&
+            if projectList.forall(e => e.deterministic && !hasSubquery(e)) && children.nonEmpty &&
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1377]: I don't think we need to check subquery here. The code in Spark 3.2 can pushdown project though union as well even if it contains subqueries.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 2]
```diff
@@ -1372,15 +1374,16 @@ object CombineUnions extends Rule[LogicalPlan] {
...
         case Project(projectList, Deduplicate(keys: Seq[Attribute], u: Union))
-            if projectList.forall(_.deterministic) && flattenDistinct && u.byName == topByName &&
-              u.allowMissingCol == topAllowMissingCol && AttributeSet(keys) == u.outputSet =>
+            if projectList.forall(e => e.deterministic && !hasSubquery(e)) && flattenDistinct &&
+              u.byName == topByName && u.allowMissingCol == topAllowMissingCol &&
+              AttributeSet(keys) == u.outputSet =>
           stack.pushAll(pushProjectionThroughUnion(projectList, u).reverse)
         case Project(projectList, u @ Union(children, byName, allowMissingCol))
-            if projectList.forall(_.deterministic) && children.nonEmpty &&
+            if projectList.forall(e => e.deterministic && !hasSubquery(e)) && children.nonEmpty &&
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1386]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 1 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: fixed

**Comment by cloud-fan** [comment]
@allisonwang-db

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 3]
```diff
@@ -1355,7 +1356,8 @@ object CombineUnions extends Rule[LogicalPlan] {
     while (stack.nonEmpty) {
       stack.pop() match {
         case p1 @ Project(_, p2: Project)
-            if canCollapseExpressions(p1.projectList, p2.projectList, alwaysInline = false) =>
+            if canCollapseExpressions(p1.projectList, p2.projectList, alwaysInline = false) &&
+              !p1.projectList.exists(hasSubquery) && !p2.projectList.exists(hasSubquery) =>
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1360]: `hasSubquery` -> `hasCorrelatedSubquery`. I think non-correlated subqueries should work but let's add a test case just to be sure.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 4]
```diff
@@ -4427,6 +4427,31 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
...
+        """
+          |SELECT x + 1
+          |FROM   (SELECT id
+          |               + (SELECT Max(id)
+          |                  FROM   range(2)) AS x
+          |        FROM   range(1)) t
+          |UNION
+          |SELECT 1 AS a
+        """.stripMargin),
+      Seq(Row(2), Row(1)))
```
Comment (wangyum) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:4453]:
```
=== Applying Rule org.apache.spark.sql.catalyst.optimizer.CombineUnions ===
 Distinct                                                          Distinct
 +- Union false, false                                             +- Union false, false
!   :- Project [(x#219L + cast(1 as bigint)) AS (x + 1)#225L]         :- Project [((id#221L + scalar-subquery#218 []) + cast(1 as bigint)) AS (x + 1)#225L]
!   :  +- Project [(id#221L + scalar-subquery#218 []) AS x#219L]      :  :  +- Aggregate [max(id#222L) AS max(id)#224L]
!   :     :  +- Aggregate [max(id#222L) AS max(id)#224L]              :  :     +- Range (0, 2, step=1, splits=None)
!   :     :     +- Range (0, 2, step=1, splits=None)                  :  +- Range (0, 1, step=1, splits=None)
!   :     +- Range (0, 1, step=1, splits=None)                        +- Project [cast(1 as bigint) AS a#226L]
!   +- Project [cast(a#220 as bigint) AS a#226L]                         +- OneRowRelation
!      +- Project [1 AS a#220]                                     
!         +- OneRowRelation    
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 3 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: +1

**Comment by dongjoon-hyun** [comment]
Could you update the PR description too, @wangyum ?

**Comment by wangyum** [comment]
Fixed.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Comment by dongjoon-hyun** [comment]
Merged to master/3.3.

cc @MaxGekk

**Pull request closed without merging**